jerry-main/fuzzilli.c

```diff
@@ -0,0 +1,300 @@
+  {
+    int fd = shm_open (shm_key, O_RDWR, S_IRUSR | S_IWUSR);
+    if (fd <= -1)
+    {
+      fprintf (stderr, "Failed to open shared memory region: %s\n", strerror (errno));
+      _exit (-1);
+    }
+
+    __shmem = (struct shmem_data *) mmap (0, SHM_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    if (__shmem == MAP_FAILED)
+    {
+      fprintf (stderr, "Failed to mmap shared memory region\n");
+      _exit (-1);
+    }
+
+    sanitizer_cov_reset_edgeguards ();
+
+    __shmem->num_edges = (uint32_t) (stop - start);
+    fprintf (stderr, "[COV] edge counters initialized. Shared memory: %s with %u edges\n", shm_key, __shmem->num_edges);
+  }
+} /* __sanitizer_cov_trace_pc_guard_init */
+
+void __sanitizer_cov_trace_pc_guard (uint32_t *guard)
+{
+  // There's a small race condition here: if this function executes in two threads for the same
+  // edge at the same time, the first thread might disable the edge (by setting the guard to zero)
+  // before the second thread fetches the guard value (and thus the index). However, our
+  // instrumentation ignores the first edge (see libcoverage.c) and so the race is unproblematic.
+  uint32_t index = *guard;
+  __shmem->edges[index / 8] |= 1 << (index % 8);
+  *guard = 0;
+} /* __sanitizer_cov_trace_pc_guard */
+
+// END FUZZING CODE
+
+// We have to assume that the fuzzer will be able to call this function e.g. by
+// enumerating the properties of the global object and eval'ing them. As such
+// this function is implemented in a way that requires passing some magic value
+// as first argument (with the idea being that the fuzzer won't be able to
+// generate this value) which then also acts as a selector for the operation
+// to perform.
+jerry_value_t
+jerryx_handler_fuzzilli (const jerry_value_t func_obj_val, /**< function object */
+                         const jerry_value_t this_p, /**< this arg */
+                         const jerry_value_t args_p[], /**< function arguments */
+                         const jerry_length_t args_cnt) /**< number of function arguments */
+{
+  (void) func_obj_val; /* unused */
+  (void) this_p;       /* unused */
+
+  if (args_cnt < 1 || !jerry_value_is_string (args_p[0]))
+  {
+    return jerry_create_error (JERRY_ERROR_TYPE, (jerry_char_t *)"Expected a string");
+  }
+
+  jerry_value_t ret_val = jerry_create_undefined ();
+  jerry_size_t str_size = jerry_get_utf8_string_size (args_p[0]);
+  jerry_char_t *operation = jerry_heap_alloc (str_size * sizeof (jerry_char_t));
+
+  if (operation == NULL || jerry_string_to_utf8_char_buffer (args_p[0], operation, str_size) != str_size)
+  {
+    return jerry_create_error (JERRY_ERROR_RANGE, (jerry_char_t *)"Internal error");
+  }
+
+  if (strcmp ((char *) operation, "FUZZILLI_CRASH") == 0)
+  {
+    if (args_cnt == 2 && jerry_value_is_number (args_p[1]))
+    {
+      int arg = (int) jerry_get_number_value (args_p[1]);
+      switch (arg)
+      {
+        case 0:
+        *((int *) 0x41414141) = 0x1337;
```

Comment:
Alright, I see that this construct is used in all targets in fuzzilli to trigger a segfault. Leave it then, but please add a comment as it is done in the duktape integration (`/* intentionally segfault */`).

(I still wonder why not `raise (SIGSEGV);` is used everywhere.)